mini-projects/01-device-driver/laptop-interface/src/controller.rs

```diff
@@ -0,0 +1,175 @@
+use std::{ffi::CString, ptr, time::Instant};
+
+use crossbeam::channel::{Receiver, Sender};
+use libc::{c_char, c_int, c_void};
+
+use crate::command::Command;
+
+#[link(name = "serialport")]
+extern "C" {
+    fn sp_get_port_by_name(name: *const c_char, port: *mut *mut c_void) -> c_int;
+    fn sp_open(port: *mut c_void, flags: c_int) -> c_int;
+    fn sp_close(port: *mut c_void) -> c_int;
+    fn sp_free_port(port: *mut c_void);
+    fn sp_blocking_read(port: *mut c_void, buf: *mut u8, count: usize, timeout_ms: u32) -> c_int;
+    fn sp_blocking_write(port: *mut c_void, buf: *const u8, count: usize, timeout_ms: u32) -> c_int;
+}
+
+#[derive(Debug, Clone, Copy)]
+struct ControllerState {
+    buttons: u8,
+    adc_value: u16,
+}
+
+fn open_port(port_name: &str) -> Result<*mut c_void, String> {
+    let port_name_c = CString::new(port_name).expect("CString::new failed");
+    let mut port: *mut c_void = ptr::null_mut();
+
+    unsafe {
+        if sp_get_port_by_name(port_name_c.as_ptr(), &mut port) != 0 {
+            return Err(format!("Failed to get port by name {}", port_name));
+        }
+        if sp_open(port, 3) != 0 {
+            sp_free_port(port);
+            return Err(format!("Failed to open port {}", port_name));
+        }
+    }
+    Ok(port)
+}
+
+fn send_command(port: *mut c_void, command: Command) -> Result<(), String> {
+    let bytes = command.as_bytes();
+    unsafe {
+        let written = sp_blocking_write(port, bytes.as_ptr(), bytes.len(), 1000);
+        if written != bytes.len() as i32 {
+            return Err(format!(
+                "{:?} failed to send {:?}: {}",
+                port, command, written
+            ));
+        }
+    }
+    Ok(())
+}
+
+fn read_controller_state(port: *mut c_void) -> Result<ControllerState, String> {
+    let mut buffer = [0u8; 3];
+    unsafe {
+        let bytes_read = sp_blocking_read(port, buffer.as_mut_ptr(), buffer.len(), 1000);
+        if bytes_read != buffer.len() as i32 {
+            return Err(format!(
+                "{:?} failed to read controller state: {}",
+                port, bytes_read
+            ));
+        }
+    }
+    let adc_value = u16::from_be_bytes([buffer[1], buffer[2]]);
+    let buttons = buffer[0];
```

Comment:
so the problem with hard-coding the indices within your buffer is, with how the firmware currently is written, the buffer is going to be more than 3 bytes (since the bit mask gets populated into a string, each digit can take up its own byte), this is likely why your button and add values appear garbage.